command/agent/agent_endpoint.go

```diff
@@ -305,6 +308,173 @@ func (s *HTTPServer) AgentMonitor(resp http.ResponseWriter, req *http.Request) (
+	followStr := req.URL.Query().Get("follow")
+	if followStr != "" {
+		parsedfollow, err := strconv.ParseBool(followStr)
+		if err != nil {
+			return nil, CodedError(400, fmt.Sprintf("Unknown option for follow %v", err))
+		}
+		follow = parsedfollow
+	}
+
+	logSince := "72" //default value
+	logSinceStr := req.URL.Query().Get("log_since")
+	if logSinceStr != "" {
+		_, err := strconv.Atoi(logSinceStr)
+		if err != nil {
+			return nil, CodedError(400, fmt.Sprintf("Unknown integer for log-since: %v", err))
+		}
+		logSince = logSinceStr
+	}
+
+	serviceName := req.URL.Query().Get("service_name")
+	nodeID := req.URL.Query().Get("node_id")
+
+	nomadLogPath := s.agent.GetConfig().LogFile
+	var mockMonitor monitor.ExportMonitor
+	if mocked := req.URL.Query().Get("mocked"); mocked != "" {
+		mockMonitor = monitor.Mock()
+	}
+	// Build the request and parse the ACL token
+	args := cstructs.MonitorExportRequest{
+		NodeID:       nodeID,
+		ServerID:     req.URL.Query().Get("server_id"),
+		LogSince:     logSince,
+		ServiceName:  serviceName,
+		OnDisk:       onDisk,
+		NomadLogPath: nomadLogPath,
+		Follow:       follow,
+		MockMonitor:  &mockMonitor,
+	}
+	if args.NodeID != "" && args.ServerID != "" {
+		return nil, CodedError(400, "Cannot target node and server simultaneously")
+	}
+
+	if args.OnDisk && args.ServiceName != "" {
+		return nil, CodedError(400, "Cannot target journalctl and nomad log file simultaneously")
+	}
+
+	if err := monitor.ScanServiceName(args.ServiceName); err != nil {
+		return nil, CodedError(422, err.Error())
+	}
+
+	if err := monitor.ScanField(args.NomadLogPath, "Nomad Log Path"); err != nil {
+		return nil, CodedError(422, err.Error())
+	}
+	// Force the Content-Type to avoid Go's http.ResponseWriter from
+	// detecting an incorrect or unsafe one.
+	resp.Header().Set("Content-Type", "application/json")
+
+	s.parse(resp, req, &args.QueryOptions.Region, &args.QueryOptions)
+
+	// Make the RPC
+	var handler structs.StreamingRpcHandler
+	var handlerErr error
+	if nodeID != "" {
+		// Determine the handler to use
+		useLocalClient, useClientRPC, useServerRPC := s.rpcHandlerForNode(nodeID)
+		if useLocalClient {
+			handler, handlerErr = s.agent.Client().StreamingRpcHandler("Agent.MonitorExport")
+		} else if useClientRPC {
+			handler, handlerErr = s.agent.Client().RemoteStreamingRpcHandler("Agent.MonitorExport")
+		} else if useServerRPC {
+			handler, handlerErr = s.agent.Server().StreamingRpcHandler("Agent.MonitorExport")
+		} else {
+			handlerErr = CodedError(400, "No local Node and node_id not provided")
+		}
+		// No node id monitor current server/client
+	} else if srv := s.agent.Server(); srv != nil {
+		handler, handlerErr = srv.StreamingRpcHandler("Agent.MonitorExport")
+	} else {
+		handler, handlerErr = s.agent.Client().StreamingRpcHandler("Agent.MonitorExport")
+	}
+
+	if handlerErr != nil {
+		return nil, CodedError(500, handlerErr.Error())
+	}
+	httpPipe, handlerPipe := net.Pipe()
```

Comment:
From here down we're sharing a _lot_ of very fussy code with the `AgentMonitor`? Can we pull out a shared function to reuse?